### PR TITLE
Started refactoring and update to python 3 features

### PR DESCRIPTION
**Pull request by flixman**
First PR of a series to refactor the project into a one more OOP based, adding unit/integration testing, annotations, logging, etc.

This PR performs some refactoring of the main file and module init files, and replaces the calls to "print" by calls to the corresponding logging module. The unit tests can be run by executing 
```python
python -m unittest
```
from the project root folder

### CODE REVIEW

**Review by grawity — COMMENTED** [review]
Altogether I'm not sure if this improves things much; the PR does a lot but without a clear focus on anything specific (and leaves the main script in a non-working state).

I like that main() is now much smaller because process_log() and compare_pcrs() are encapsulated in functions and are clearly separate from each other, but they're still entangled with `args` just as before. Rather than trying to pull ArgumentParser into tests, it would be better to have tests that are able to call process_log() without going through argparse at all.

(I'd actually call this a "large" pull – even though it has only a few commits, each of them makes a lot of unrelated changes, which makes both the individual commits and the PR as a whole harder to review; e.g. adding tests should be a separate commit from moving code around, and some things should probably be whole separate PRs to make them easier to discuss.)

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 /*.egg-info
 /build
 /dist
+/.idea/
```
[the remark this answers is not in the document]
Comment (flixman) [reply .gitignore:5]: Indeed, I introduced it out of habit.

File: tpm_futurepcr.py [hunk 2]
```diff
@@ -1,3 +1,39 @@
...
+    possibly_unused_bank
+import tpm_futurepcr.logging as logging
+
+logging.basicConfig(stream=sys.stdout)
+logger = logging.getLogger('tpm_futurepcr')
+
+
+if __name__ == "__main__":
+    parser = create_argparser()
+    args = postprocess_args(parser.parse_args())
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr.py:14]: Agree. Moved it to the main file, and the library methods now depend on single parameters (so, not on argparse namespaces).

File: tpm_futurepcr/__init__.py [hunk 3]
```diff
@@ -12,81 +12,93 @@
...
+            if not isinstance(values, list):
+                values = [values]
+            kvpairs = [v.split("=", 1) for v in values]
+            try:
+                kvmap.update(kvpairs)
+            except ValueError:
+                raise argparse.ArgumentTypeError("Value for option %s malformed: %s" % (self.dest, values))
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-L", "--pcr-list", type=_validate_pcrlist, default=','.join(list(map(str, range(24)))), help="comma-separated list of PCR registers")
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/__init__.py:48]: The reason for the validate_pcrlist and the postprocess_args is that hash_alg can be specified in different locations. On the validate_pcrlist I have added value checks, so I think is good to have it around... but I agree that the solution is not very elegant. Should you be completely against it, just say it openly and will remove it.

File: tpm_futurepcr/util.py [hunk 4]
```diff
@@ -15,15 +15,17 @@ def to_hex(buf):
...
-    else:
-        max_len = min(max_len, len(buf))
+    # max_len must be smaller than len(buf), if defined
+    max_len = min(max_len or len(buf), len(buf))
+
+    # print the hex codes and their ascii representation
     for i in range(0, max_len, 16):
         row = buf[i:i+16]
-        hexs = ["%02X" % b for b in row] + ["  "] * 16
-        text = [chr(b) if 0x20 < b < 0x7f else "." for b in row] + [" "] * 16
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/util.py:20]: Indeed! Sorry xD.

[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/util.py:20]: Right, sorry. Then, what I propose is to initialize hexs and text, and overwrite the elements for which row has a content.

File: tpm_futurepcr/util.py [hunk 5]
```diff
@@ -2,37 +2,47 @@
...
-        print(offs, " ".join(hexs[:16]), "|%s|" % "".join(text[:16]))
+        hexs = ["%02X" % b for b in row]
+        text = [chr(b) if 0x20 < b < 0x7f else "." for b in row]
+        logger.debug("0x%08x: %s |%s|", i, " ".join(hexs[:16]), "".join(text[:16]))
+
+    # notify the user in case there were bytes left unprinted
     if len(buf) > max_len:
-        print("(%d more bytes)" % (len(buf) - max_len))
+        logger.debug("(%d more bytes)", len(buf) - max_len)
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/util.py:30]: understood, I have moved it outwards.

File: tpm_futurepcr/event_log.py [hunk 6]
```diff
@@ -1,9 +1,13 @@
-from pprint import pprint
+from pprint import pformat
 
-from .binary_reader import BinaryReader
 from .device_path import *
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/event_log.py:3]: Indeed, only parse_efi_device_path seems to be required in event_log.

File: tpm_futurepcr/__init__.py [hunk 7]
```diff
@@ -12,81 +12,93 @@
...
+    return parser
+
+
+def postprocess_args(args: argparse.Namespace) -> argparse.Namespace:
+    if (args.hash_alg is None and args.pcr_list[1] is None) or \
+       (args.hash_alg is not None and args.pcr_list[1] is not None and args.hash_alg != args.pcr_list[1]):
+        raise argparse.ArgumentTypeError("PCR hash algorithm must be explicitly specified either in the pcr list or with the -H flag")
+
+    # populate properly the hash_alg argument
+    args.hash_alg = TpmAlgorithm[args.hash_alg or args.pcr_list[1].upper()]
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/__init__.py:68]: Indeed. Fixed. I planned to add more tests for the argparse and so in the following PRs.

File: tpm_futurepcr/__init__.py [hunk 8]
```diff
@@ -12,81 +12,93 @@
...
             show_log_entry(event)
 
         if idx == 0xFFFFFFFF:
             if _verbose_pcr:
-                print("event updates Windows virtual PCR[-1], skipping")
+                logger.verbose("event updates Windows virtual PCR[-1], skipping")
             continue
 
-        this_extend_value = event["pcr_extend_values"].get(tpm_hash_alg)
+        this_extend_value = event["pcr_extend_values"].get(args.hash_alg)
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/__init__.py:96]: This has been fixed as part of a previous comment.

File: tpm_futurepcr.py [hunk 9]
```diff
@@ -21,21 +23,21 @@
     for event in enum_log_entries(args.log_path):
         idx = event["pcr_idx"]
 
-        _verbose_pcr = (args.verbose and (verbose_all_pcrs or idx in wanted_pcrs))
+        _verbose_pcr = logger.level <= logging.VERBOSE and idx in wanted_pcrs
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr.py]: I overlooked this, right. This means that when a user specifies all the indexes (or leaves the argument at its default), the system does not report information about those indexes specified only... Don´t you think this is worth a parameter, like "verbose-exotic-pcrs" or so?

**Comment by flixman** [comment]
Thank you for the review, grawity. With the exception of one of your comments, for which I have posed a question, I think I have implemented the remaining. This PR is not supposed to improve the code a lot, but to start encapsulating code, bringing unit/integration testing, etc. I will make the next ones smaller.

File: tpm_futurepcr/__init__.py [hunk 10]
```diff
@@ -12,81 +12,93 @@
...
+    parser.add_argument("-H", "--hash-alg", help="specify the hash algorithm", choices=['sha1', 'sha256'])
+    parser.add_argument("-o", "--output", type=Path, help="write binary PCR values to specified file")
+    parser.add_argument("--allow-unexpected-bsa", action="store_true", help="accept BOOT_SERVICES_APPLICATION events with weird paths")
+    parser.add_argument("--substitute-bsa-unix-path", action=_KeyValueAction, help="substitute BOOT_SERVICES_APPLICATION path (syntax: <computed unix path>=<new unix path>)")
+    parser.add_argument("--compare", action="store_true", help="compare computed PCRs against live values")
+    parser.add_argument("--log-path", type=Path, help="read binary log from an alternative path")
+
+    parser.add_argument('-d', '--debug', help="Print lots of debugging statements", action="store_const", dest="loglevel", const=logging.DEBUG, default=logging.INFO)
+    parser.add_argument('-v', '--verbose', help="Be verbose", action="store_const", dest="loglevel", const=logging.VERBOSE)
+
```
[the remark this answers is not in the document]
Comment (flixman) [reply tpm_futurepcr/__init__.py]: I was under the assumption that "verbose" is for the user, whereas "debug" is for the developers? This means that a user can request being "verbose" meaning, e.g., "I want to know when device paths are detected" whereas a developer might specify "debug" meaning "I want to know the contents of all the events".

File: tpm_futurepcr.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (grawity) [reply tpm_futurepcr.py]: Hmm, for now I think the functionality can just go away.

File: tpm_futurepcr/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (grawity) [reply tpm_futurepcr/__init__.py]: Yeah, a function to parse and validate the PCR list string is useful, but on the other hand I don't see the value of postprocess_args (i.e. trying to update `args` with the parsed values) vs using local variables. Is the goal to be able to reuse this logic in tests, or is it to be able to use `args.foo` in the remaining code?

**Comment by flixman** [comment]
@grawity did you receive my last message?

**Comment by grawity** [comment]
I think I'll just merge this for now... what goes into my todo list stays in the list for months.

**Comment by grawity** [comment]
While using the new version, I was reminded why verbose_all_pcrs actually exists: without it, `-L 12` doesn't recognize the systemd-boot EV_IPL because it doesn't know the `last_efi_binary` from PCR 4. Previously it would still have processed PCR 4 and recorded the path even if the PCR wasn't in the requested list. (In other words, there can be cross-PCR dependencies.)

I'm also getting a:

```
$ sudo ./tpm_futurepcr.py -v -d -L sha256:4,12
VERBOSE:event_log:PCR 4 -- Event 00000004 <TpmEventType.SEPARATOR>
DEBUG:event_log:0x00000000: 00 00 00 00                                     |....                        |
VERBOSE:tpm_futurepcr:--> after this event, PCR 4 contains value 3d458cfe55cc03ea1f443f1562beec8df51c75e14a9fcf9a7234a13f198e7969
VERBOSE:tpm_futurepcr:--> after reboot, PCR 4 will contain value 3d458cfe55cc03ea1f443f1562beec8df51c75e14a9fcf9a7234a13f198e7969
VERBOSE:event_log:PCR 4 -- Event 80000003 <TpmEventType.EFI_BOOT_SERVICES_APPLICATION>
VERBOSE:event_log:Path vector:
VERBOSE:event_log:  * ACPIDevice           ACPI                 b'\xd0A\x03\n\x00\x00\x00\x00'
VERBOSE:event_log:  * HardwareDevice       PCI                  b'\x00\x1b'
VERBOSE:event_log:  * HardwareDevice       PCI                  b'\x00\x00'
VERBOSE:event_log:  * MessagingDevice      NVMe                 b'\x01\x00\x00\x00\x00&\xb7h][Q%'
VERBOSE:event_log:  * MediaDevice          HardDrive            b'\x01\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00>6\x0evo\xc1AL\xb0\xaf\x05\x9a\x1d+\x94#\x02\x02'
VERBOSE:event_log:  * MediaDevice          FilePath             \EFI\systemd\systemd-bootx64.efi
VERBOSE:event_log:  * End                  255                  b''
Traceback (most recent call last):
  File "/home/grawity/src/tpm_futurepcr/./tpm_futurepcr.py", line 94, in <module>
    this_pcrs, next_pcrs, errors = process_log(args.pcr_list, args.hash_alg, args.log_path, args.substitute_bsa_unix_path, args.allow_unexpected_bsa)
  File "/home/grawity/src/tpm_futurepcr/tpm_futurepcr/__init__.py", line 56, in process_log
    file_hash = hash_pecoff(unix_path, hash_alg)
  File "/home/grawity/src/tpm_futurepcr/tpm_futurepcr/util.py", line 65, in hash_pecoff
    fpr.add_authenticode_hashers(getattr(hashlib, alg))
TypeError: getattr(): attribute name must be string
```

**Comment by flixman** [comment]
@grawity I will fix that on my next commit :-) I do not have a TPM right now, so I cannot reproduce this error you are getting. Is there any way you can anonymize your tpm event log and upload it to the repository, so that I can use it for testing?

**Pull request merged**